pkg/client/client_test.go

```diff
@@ -1,56 +1,660 @@
+						Namespace: "bar",
+					},
+				},
+				namespace: "bar",
+			},
+		},
+	}
+	for _, tt := range tests {
+		test := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			mockRESTClient := &fake.RESTClient{
+				GroupVersion:         test.fields.GVR.GroupVersion(),
+				NegotiatedSerializer: negotiatedSerializer,
+			}
+			c := NewClient(test.fields.GVR, test.fields.Kind, test.fields.Namespaced, mockRESTClient, 0)
+			handler := newRequestHandler(c, test.args.obj, test.args.namespace, false, false)
+			mockRESTClient.Client = fake.CreateHTTPClient(handler)
+			err := c.Delete(context.TODO(), test.args.namespace, test.args.obj.GetName(), metav1.DeleteOptions{})
+			if test.wantErr {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+		})
+	}
+}
 
-	_, err = testRestClientWithAgent.Client.Transport.RoundTrip(&http.Request{})
-	assert.Nil(t, err)
+func TestClient_DeleteCollection(t *testing.T) {
+	t.Parallel()
+	tests := []*testInfo{
+		{
+			name: "base test",
+			fields: clientFields{
+				GVR: schema.GroupVersionResource{
+					Group:    "",
+					Version:  "v1",
+					Resource: "pods",
+				},
+				Namespaced: true,
+				Kind:       "Pod",
+			},
+			args: testArgs{
+				namespace: "bar",
+			},
+		},
+	}
+	for _, tt := range tests {
+		test := tt
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			mockRESTClient := &fake.RESTClient{
+				GroupVersion:         test.fields.GVR.GroupVersion(),
+				NegotiatedSerializer: negotiatedSerializer,
+			}
+			c := NewClient(test.fields.GVR, test.fields.Kind, test.fields.Namespaced, mockRESTClient, 0)
+			handler := newRequestHandler(c, test.args.desired, test.args.namespace, true, false)
+			mockRESTClient.Client = fake.CreateHTTPClient(handler)
+			err := c.DeleteCollection(context.TODO(), test.args.namespace, metav1.DeleteOptions{}, metav1.ListOptions{})
+			if test.wantErr {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+		})
+	}
+}
 
-	// with invalid config
-	testClient = &Client{
-		Config: rest.Config{},
+func TestClient_Watch(t *testing.T) {
+	t.Parallel()
+	tests := []*testInfo{
+		{
+			name: "base test",
+			fields: clientFields{
+				GVR: schema.GroupVersionResource{
+					Group:    "",
+					Version:  "v1",
+					Resource: "pods",
+				},
+				Namespaced: true,
+				Kind:       "Pod",
+			},
+			args: testArgs{
+				namespace: "bar",
+			},
+		},
 	}
+	for _, tt := range tests {
+		test := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			mockRESTClient := &fake.RESTClient{
+				GroupVersion:         test.fields.GVR.GroupVersion(),
+				NegotiatedSerializer: negotiatedSerializer,
+			}
+			c := NewClient(test.fields.GVR, test.fields.Kind, test.fields.Namespaced, mockRESTClient, 0)
+			handler := newRequestHandler(c, test.args.desired, test.args.namespace, true, false)
+			mockRESTClient.Client = fake.CreateHTTPClient(handler)
+			_, err := c.Watch(context.TODO(), test.args.namespace, metav1.ListOptions{})
```

Comment:
We wouldn't be testing anything in lasso at that point if we are making a fake event interface and verifying that our fake interface received our fake data. Testing of the interface itself should be left up to the package that implements the interface, not the one using it.

Reply:
I'm not sure that I agree. We would want to test that lasso returned an interface with the type/structure that we expected. That's tough here since this is returning an interface, but at minimum I would think we would want to make sure it's non-nil in non-error case.

Reply:
> We would want to test that lasso returned an interface with the structure that we expected

We do not know what struct to expect, nor do we care that functionality is left up to K8s. We can't assume that the interface should be non-nil when a non-error is returned because K8s determines that behavior. This means any test around that behavior would just be testing an external package.

Reply:
Discussed offline. Since this is handed off directly to k8s without us modifying/processing the return in anyway, it will be difficult to test this without testing the underlying library, which we want to avoid. Therefore, we are going to leave a comment indicating that the return values are entirely controlled by the underlying k8s library. Resolving.